src/uu/tail/src/chunks.rs

```diff
@@ -81,3 +91,630 @@ impl<'a> Iterator for ReverseChunks<'a> {
+
+        Ok(Some(self.bytes))
+    }
+}
+
+/// An abstraction layer on top of [`BytesChunk`] mainly to simplify filling only the needed amount
+/// of chunks. See also [`Self::fill`].
+pub(crate) struct BytesChunkBuffer {
+    /// The number of bytes to print
+    num_print: u64,
+    /// The current number of bytes summed over all stored chunks in [`Self::chunks`]
+    bytes: u64,
+    /// The buffer to store [`BytesChunk`] in
+    chunks: VecDeque<Box<BytesChunk>>,
+}
+
+impl BytesChunkBuffer {
+    /// Creates a new [`BytesChunkBuffer`].
+    ///
+    /// # Arguments
+    ///
+    /// * `num_print`: The number of bytes to print
+    ///
+    /// returns: BytesChunk
+    ///
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    /// let mut chunk = BytesChunk::new();
+    /// chunk.buffer[1] = 1;
+    /// chunk.bytes = 2;
+    /// let new_chunk = BytesChunk::from_chunk(&chunk, 0);
+    /// assert_eq!(2, new_chunk.get_buffer().len());
+    /// assert_eq!(&[0, 1], new_chunk.get_buffer());
+    ///
+    /// let new_chunk = BytesChunk::from_chunk(&chunk, 1);
+    /// assert_eq!(1, new_chunk.get_buffer().len());
+    /// assert_eq!(&[1], new_chunk.get_buffer());
+    /// ```
+    pub(crate) fn new(num_print: u64) -> Self {
+        Self {
+            bytes: 0,
+            num_print,
+            chunks: VecDeque::new(),
+        }
+    }
+
+    /// Fills this buffer with chunks and consumes the reader completely. This method ensures that
+    /// there are exactly as many chunks as needed to match [`Self::num_print`] bytes, so there are
+    /// in sum exactly `self.num_print` bytes stored in all chunks. The method returns an iterator
+    /// over these chunks. If there are no chunks, for example because the piped stdin contained no
+    /// bytes, or `num_print = 0` then `iterator.next` returns None.
+    ///
+    /// # Arguments
+    ///
+    /// * `reader`: A [`BufReader`] with an inner element implementing the [`Read`] trait.
+    ///
+    /// returns: Result<Iter<Box<BytesChunk, Global>>, Box<dyn UError, Global>>
+    ///
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    /// use std::io::{BufReader, Cursor};
+    ///
+    /// let mut reader = BufReader::new(Cursor::new(""));
+    /// let num_print = 0;
+    /// let mut chunks = BytesChunkBuffer::new(num_print);
+    /// let mut iter = chunks.fill(&mut reader).unwrap();
+    ///
+    /// let chunk = iter.next();
+    /// assert!(chunk.is_none());
+    ///
+    /// let mut reader = BufReader::new(Cursor::new("a"));
+    /// let num_print = 1;
+    /// let mut chunks = BytesChunkBuffer::new(num_print);
+    /// let mut iter = chunks.fill(&mut reader).unwrap();
+    ///
+    /// let chunk = iter.next();
+    /// assert!(chunk.is_some());
+    /// assert_eq!(&[b'a'], chunk.unwrap().get_buffer());
+    /// assert_eq!(None, iter.next());
+    /// ```
+    pub(crate) fn fill(
+        &mut self,
+        reader: &mut BufReader<impl Read>,
+    ) -> UResult<Iter<Box<BytesChunk>>> {
```

Comment:
I wonder if this might be cleaner if it doesn't return the iterator right away? I.e. this would just return `UResult<()>` and then you can just get the `VecDeque` via `self.chucks` after and iterate over it.

Reply:
Or you could have the logic for printing all chucks that's now in `tail.rs` in a method and then you don't need to return the chunks at all.

Reply:
I like the idea of having a printing function. Concerning the return type, I actually like having something useful returned if I can. The caller can still decide to ignore it. Would returning the total number of lines or total number of bytes contained in all chunks be an alternative for you?

Reply:
I understand what you mean, but since were writing an internal function with a very well-defined goal, I think we should limit it to what we need and not add any unnecessary fluff. If the total number of lines is actually used somewhere then it's good, otherwise I'd prefer to leave it out.